pkg/services/ngalert/migration/channel.go

```diff
@@ -189,197 +103,58 @@ func (om *OrgMigration) createNotifier(c *legacymodels.AlertNotification) (*apim
-		for _, c := range defaultChannels {
-			// Need to create a new notifier to prevent uid conflict.
-			defaultNotifier, err := om.createNotifier(c)
-			if err != nil {
-				return nil, nil, err
-			}
-
-			newDefaultReceiver.GrafanaManagedReceivers = append(newDefaultReceiver.GrafanaManagedReceivers, defaultNotifier)
-
-			// Choose the lowest send reminder duration from all the notifiers to use for default route.
-			if c.SendReminder && c.Frequency < time.Duration(repeatInterval) {
-				repeatInterval = model.Duration(c.Frequency)
-			}
-		}
-	} else {
-		// If there is only a single default channel, we don't need a separate receiver to hold it. We can reuse the existing receiver for that single notifier.
-		defaultRoute.Receiver = defaultChannels[0].Name
-		if defaultChannels[0].SendReminder {
-			repeatInterval = model.Duration(defaultChannels[0].Frequency)
-		}
-
-		// No need to create a new receiver.
-		newDefaultReceiver = nil
-	}
-	defaultRoute.RepeatInterval = &repeatInterval
-
-	return newDefaultReceiver, defaultRoute, nil
+	}, nil
 }
 
-// Create one route per contact point, matching based on ContactLabel.
-func createRoute(cr channelReceiver) (*apimodels.Route, error) {
-	// We create a regex matcher so that each alert rule need only have a single ContactLabel entry for all contact points it sends to.
-	// For example, if an alert needs to send to contact1 and contact2 it will have ContactLabel=`"contact1","contact2"` and will match both routes looking
-	// for `.*"contact1".*` and `.*"contact2".*`.
-
-	// We quote and escape here to ensure the regex will correctly match the ContactLabel on the alerts.
-	name := fmt.Sprintf(`.*%s.*`, regexp.QuoteMeta(quote(cr.receiver.Name)))
-	mat, err := labels.NewMatcher(labels.MatchRegexp, ContactLabel, name)
-	if err != nil {
-		return nil, err
+// createRoute creates a route from a legacy notification channel, and matches using a label based on the channel UID.
+func createRoute(channel *legacymodels.AlertNotification, receiverName string) *apimodels.Route {
+	// We create a matchers based on channel UID so that we only need a single route per channel.
+	// All routes are stored in a nested route under the root. This is so we can keep the migrated channels separate
+	// and organized.
+	// The matchers are created using a label that is unique to the channel UID. So, each migrated alert rule can define
+	// one label per migrated channel that it should send to.
+	// Default channels are matched using a catch-all matcher, because in legacy alerting they are attached to all
+	// alerts automatically.
+	//
+	// For example, if an alert needs to send to channel1 and channel2 it will have two labels:
+	// - __contact_channel1__="true"
+	// - __contact_channel2__="true"
+	//
+	// These will match two routes as they are all defined with Continue=true.
+
+	label := fmt.Sprintf(ContactLabelTemplate, channel.UID)
+	mat, _ := labels.NewMatcher(labels.MatchEqual, label, "true")
```

Comment:
`labels` is an external dependency and it's implementation details can change in future. I think it is a good practice to correctly handle the function result, in this case, error.

Reply:
That's fair, I went back and forth on this one. I'll add it back.